agent/hcp/telemetry_provider_test.go

```diff
@@ -0,0 +1,341 @@
+			},
+			expected: &testConfig{
+				endpoint: "http://test.com/v1/metrics",
+				labels: map[string]string{
+					"test_label": "123",
+				},
+				filters:         "test",
+				refreshInterval: testRefreshInterval,
+			},
+			metricKey: testMetricKeyFailure,
+		},
+		"sameConfigHCPClientFailure": {
+			optsInputs: &testConfig{
+				endpoint: "http://test.com/v1/metrics",
+				filters:  "test",
+				labels: map[string]string{
+					"test_label": "123",
+				},
+				refreshInterval: testRefreshInterval,
+			},
+			mockExpect: func(m *client.MockClient) {
+				m.EXPECT().FetchTelemetryConfig(mock.Anything).Return(nil, fmt.Errorf("failure"))
+			},
+			expected: &testConfig{
+				endpoint: "http://test.com/v1/metrics",
+				filters:  "test",
+				labels: map[string]string{
+					"test_label": "123",
+				},
+				refreshInterval: testRefreshInterval,
+			},
+			metricKey: testMetricKeyFailure,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			sink := initGlobalSink()
+			mockClient := client.NewMockClient(t)
+			tc.mockExpect(mockClient)
+
+			dynamicCfg, err := testDynamicCfg(tc.optsInputs)
+			require.NoError(t, err)
+
+			provider := &hcpProviderImpl{
+				hcpClient: mockClient,
+				cfg:       dynamicCfg,
+			}
+
+			provider.getUpdate(context.Background())
+
+			// Verify endpoint provider returns correct config values.
+			require.Equal(t, tc.expected.endpoint, provider.GetEndpoint().String())
+			require.Equal(t, tc.expected.filters, provider.GetFilters().String())
+			require.Equal(t, tc.expected.labels, provider.GetLabels())
+
+			// Verify count for transform success metric.
+			interval := sink.Data()[0]
+			require.NotNil(t, interval, 1)
+			sv := interval.Counters[tc.metricKey]
+			assert.NotNil(t, sv.AggregateSample)
+			require.Equal(t, sv.AggregateSample.Count, 1)
+		})
+	}
+}
+
+func TestTelemetryConfigProvider_Race(t *testing.T) {
+	cfg, err := testTelemetryCfg(&testConfig{
+		endpoint: "http://test.com/v1/metrics",
+		filters:  "test",
+		labels: map[string]string{
+			"test_label": "123",
+		},
+		refreshInterval: testRefreshInterval,
+	})
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	mockClient := client.NewMockClient(t)
+	mockClient.EXPECT().FetchTelemetryConfig(mock.Anything).Return(cfg, nil)
+
+	// Start the provider goroutine
+	// Every refresh interval, config will be modified.
+	provider, err := NewHCPProvider(ctx, mockClient, cfg)
+	require.NoError(t, err)
+
+	// Every refresh interval, try to query config using Get* methods inducing a race condition.
+	timer := time.NewTimer(testRefreshInterval)
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			wg := &sync.WaitGroup{}
+			// Start goroutines that try to access label configuration.
+			kickOff(wg, testRaceSampleCount, provider, func(provider *hcpProviderImpl) {
+				require.Equal(t, provider.GetLabels(), cfg.MetricsConfig.Labels)
```

Comment:
Based in our Friday meeting, we decided to let unit tests guarantee behaviour of the state change, and not check the ticker. The `checkUpdate` unit test should guarantee config changes, and this test is just a lock test, that's correct.

The issue is we have then play around with some timing stuff? How do we know the expected updated field? If we initiate an update, we'd have to make sure it's updated once (playing around with timings basically we'd need some sort of mechanism without having direct access to the channels). 

This would have been easy to do with the previous version of the code (with channel access) but I'm not sure if it would be more complex now (manage timings properly etc). If you have any easy ideas, we could pair?


Wonder what

Reply:
I agree it's not the ideal tho, I pivoted mainly based on our last discussion

Reply:
Im not really concerned with the channel communication but it seems to me like this isn't really testing much. If a single config refresh was successful all of these tests will pass. So if anyone changes your logic (ie changes the refresh logic) how do you guarantee that the readers get the new values. It appears to me at a quick glance that due to the initial load the test would never actually catch if refresh behaviour is changed or running. Do you agree?  This might be okay if you've already tested that in other test scenarios but then I don't really know what the point of this test is? Are you testing a race condition with a race flag?